File: android/src/main/kotlin/com/phonetechbd/star_printer_plus/StarPrinterPlusPlugin.kt
package com.phonetechbd.star_printer_plus

import android.content.Context
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.EventChannel
import kotlinx.coroutines.*
import com.starmicronics.stario10.*

class FlutterStarPrinterPlugin: FlutterPlugin, MethodChannel.MethodCallHandler, EventChannel.StreamHandler {
    private lateinit var channel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private lateinit var context: Context
    private var printers = mutableMapOf<String, StarPrinter>()
    private var eventSink: EventChannel.EventSink? = null

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "flutter_star_prnt")
        channel.setMethodCallHandler(this)
        
        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "flutter_star_prnt_events")
        eventChannel.setStreamHandler(this)
        
        context = flutterPluginBinding.applicationContext
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        when (call.method) {
            "portDiscovery" -> handlePortDiscovery(call, result)
            "connect" -> handleConnect(call, result)
            "disconnect" -> handleDisconnect(call, result)
            "print" -> handlePrint(call, result)
            "checkStatus" -> handleCheckStatus(call, result)
            else -> result.notImplemented()
        }
    }

    private fun handlePortDiscovery(call: MethodCall, result: Result) {
        val type = call.argument<String>("type") ?: return result.error("INVALID_ARGUMENT", "Type is required", null)
        
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val interfaceType = when(type.uppercase()) {
                    "ALL" -> InterfaceType.All
                    "LAN" -> InterfaceType.Lan
                    "BLUETOOTH" -> InterfaceType.Bluetooth
                    "USB" -> InterfaceType.Usb
                    else -> throw Exception("Unsupported interface type")
                }

                val printers = StarDeviceDiscoveryManager.discover(interfaceType, context)
                val printerList = printers.map { printer ->
                    mapOf(
                        "portName" to printer.identifier,
                        "macAddress" to printer.macAddress,
                        "modelName" to printer.information?.model?.name,
                        "USBSerialNumber" to printer.usbSerialNumber
                    )
                }
                
                withContext(Dispatchers.Main) {
                    result.success(printerList)
                }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    result.error("DISCOVERY_ERROR", e.message, null)
                }
            }
        }
    }

    private fun handleConnect(call: MethodCall, result: Result) {
        val portName = call.argument<String>("portName") ?: return result.error("INVALID_ARGUMENT", "Port name is required", null)
        val emulation = call.argument<String>("emulation") ?: return result.error("INVALID_ARGUMENT", "Emulation is required", null)
        
        CoroutineScope(Dispatchers.IO).launch {
            try {
                if (printers.containsKey(portName)) {
                    throw Exception("Printer already connected")
                }

                val settings = StarConnectionSettings(detectInterfaceType(portName), portName)
                val printer = StarPrinter(settings, context)
                
                printer.printerDelegate = createPrinterDelegate(portName)
                printer.openAsync().await()
                
                printers[portName] = printer
                
                withContext(Dispatchers.Main) {
                    result.success(null)
                }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    result.error("CONNECTION_ERROR", e.message, null)
                }
            }
        }
    }

    private fun handleDisconnect(call: MethodCall, result: Result) {
        val portName = call.argument<String>("portName") ?: return result.error("INVALID_ARGUMENT", "Port name is required", null)
        
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val printer = printers[portName] ?: throw Exception("Printer not connected")
                printer.closeAsync().await()
                printers.remove(portName)
                
                withContext(Dispatchers.Main) {
                    result.success(null)
                }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    result.error("DISCONNECTION_ERROR", e.message, null)
                }
            }
        }
    }

    private fun handlePrint(call: MethodCall, result: Result) {
        val portName = call.argument<String>("portName") ?: return result.error("INVALID_ARGUMENT", "Port name is required", null)
        val emulation = call.argument<String>("emulation") ?: return result.error("INVALID_ARGUMENT", "Emulation is required", null)
        val commands = call.argument<List<Map<String, Any>>>("printCommands") ?: return result.error("INVALID_ARGUMENT", "Print commands are required", null)
        
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val printer = printers[portName] ?: throw Exception("Printer not connected")
                val builder = StarXpandCommand.Builder()
                
                builder.addDocument(createDocumentBuilder(commands))
                
                val printCommands = builder.getCommands()
                printer.printAsync(printCommands).await()
                
                val status = printer.getStatusAsync().await()
                val statusMap = createStatusMap(status)
                
                withContext(Dispatchers.Main) {
                    result.success(statusMap)
                }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    result.error("PRINT_ERROR", e.message, null)
                }
            }
        }
    }

    private fun handleCheckStatus(call: MethodCall, result: Result) {
        val portName = call.argument<String>("portName") ?: return result.error("INVALID_ARGUMENT", "Port name is required", null)
        
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val printer = printers[portName] ?: throw Exception("Printer not connected")
                val status = printer.getStatusAsync().await()
                val statusMap = createStatusMap(status)
                
                withContext(Dispatchers.Main) {
                    result.success(statusMap)
                }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    result.error("STATUS_ERROR", e.message, null)
                }
            }
        }
    }

    private fun createDocumentBuilder(commands: List<Map<String, Any>>): StarXpandCommand.DocumentBuilder {
        val documentBuilder = StarXpandCommand.DocumentBuilder()
        val printerBuilder = StarXpandCommand.PrinterBuilder()

        commands.forEach { command ->
            when {
                command.containsKey("appendEncoding") -> {
                    // Handle encoding
                }
                command.containsKey("appendCutPaper") -> {
                    val cutType = when(command["appendCutPaper"] as String) {
                        "FullCut" -> StarXpandCommand.CutType.Full
                        "PartialCut" -> StarXpandCommand.CutType.Partial
                        else -> StarXpandCommand.CutType.Full
                    }
                    printerBuilder.addCut(cutType)
                }
                command.containsKey("openCashDrawer") -> {
                    printerBuilder.addDrawer(StarXpandCommand.DrawerChannel.No1)
                }
                command.containsKey("appendBitmap") -> {
                    // Handle bitmap
                }
                command.containsKey("appendBitmapText") -> {
                    val text = command["appendBitmapText"] as String
                    printerBuilder.addText(text)
                }
                // Add other command handlers
            }
        }

        return documentBuilder.addPrinter(printerBuilder)
    }

    private fun createStatusMap(status: StarPrinterStatus): Map<String, Any> {
        return mapOf(
            "offline" to !status.isOnline,
            "coverOpen" to status.coverOpen,
            "cutterError" to status.cutterError,
            "receiptPaperEmpty" to status.paperEmpty,
            "overTemp" to status.overTemp,
            "isSuccess" to status.isOnline
        )
    }

    private fun createPrinterDelegate(portName: String): PrinterDelegate {
        return object : PrinterDelegate() {
            override fun onStatusChanged(printer: StarPrinter) {
                super.onStatusChanged(printer)
                
                CoroutineScope(Dispatchers.IO).launch {
                    try {
                        val status = printer.getStatusAsync().await()
                        val statusMap = createStatusMap(status)
                        
                        withContext(Dispatchers.Main) {
                            eventSink?.success(statusMap)
                        }
                    } catch (e: Exception) {
                        withContext(Dispatchers.Main) {
                            eventSink?.error("STATUS_ERROR", e.message, null)
                        }
                    }
                }
            }
        }
    }

    private fun detectInterfaceType(portName: String): InterfaceType {
        return when {
            portName.startsWith("TCP:") -> InterfaceType.Lan
            portName.startsWith("BT:") -> InterfaceType.Bluetooth
            portName.startsWith("USB:") -> InterfaceType.Usb
            else -> throw Exception("Unknown interface type")
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
    }
}

File: ios/Classes/StarPrinterPlusPlugin.swift
import Flutter
import UIKit
import StarIO10

public class FlutterStarPrinterPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    private var printers: [String: StarPrinter] = [:]
    private var eventSink: FlutterEventSink?
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "flutter_star_prnt", binaryMessenger: registrar.messenger())
        let eventChannel = FlutterEventChannel(name: "flutter_star_prnt_events", binaryMessenger: registrar.messenger())
        
        let instance = FlutterStarPrinterPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
        eventChannel.setStreamHandler(instance)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "portDiscovery":
            handlePortDiscovery(call, result: result)
        case "connect":
            handleConnect(call, result: result)
        case "disconnect":
            handleDisconnect(call, result: result)
        case "print":
            handlePrint(call, result: result)
        case "checkStatus":
            handleCheckStatus(call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func handlePortDiscovery(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any],
              let typeString = arguments["type"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Type is required", details: nil))
            return
        }
        
        Task {
            do {
                let interfaceType = try getInterfaceType(from: typeString)
                let printers = try await StarDeviceDiscoveryManager.discover(interfaceType)
                
                let printerList = printers.map { printer -> [String: Any] in
                    return [
                        "portName": printer.identifier,
                        "macAddress": printer.macAddress ?? "",
                        "modelName": printer.information?.model.name ?? "",
                        "USBSerialNumber": printer.usbSerialNumber ?? ""
                    ]
                }
                
                DispatchQueue.main.async {
                    result(printerList)
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(code: "DISCOVERY_ERROR", message: error.localizedDescription, details: nil))
                }
            }
        }
    }
    
    private func handleConnect(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any],
              let portName = arguments["portName"] as? String,
              let emulation = arguments["emulation"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Port name and emulation are required", details: nil))
            return
        }
        
        Task {
            do {
                if printers[portName] != nil {
                    throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Printer already connected"])
                }
                
                let interfaceType = try detectInterfaceType(from: portName)
                let settings = StarConnectionSettings(interfaceType: interfaceType, identifier: portName)
                let printer = StarPrinter(settings)
                
                printer.printerDelegate = self
                try await printer.open()
                
                printers[portName] = printer
                
                DispatchQueue.main.async {
                    result(nil)
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(code: "CONNECTION_ERROR", message: error.localizedDescription, details: nil))
                }
            }
        }
    }
    
    private func handleDisconnect(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any],
              let portName = arguments["portName"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Port name is required", details: nil))
            return
        }
        
        Task {
            do {
                guard let printer = printers[portName] else {
                    throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Printer not connected"])
                }
                await printer.close()
                printers.removeValue(forKey: portName)
                
                DispatchQueue.main.async {
                    result(nil)
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(code: "DISCONNECTION_ERROR", message: error.localizedDescription, details: nil))
                }
            }
        }
    }
    
    private func handlePrint(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any],
              let portName = arguments["portName"] as? String,
              let emulation = arguments["emulation"] as? String,
              let commands = arguments["printCommands"] as? [[String: Any]] else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Invalid arguments", details: nil))
            return
        }
        
        Task {
            do {
                guard let printer = printers[portName] else {
                    throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Printer not connected"])
                }
                
                let builder = StarXpandCommand.Builder()
                builder.addDocument(createDocumentBuilder(from: commands))
                
                let printCommands = try builder.getCommands()
                try await printer.print(printCommands)
                
                let status = try await printer.getStatus()
                let statusMap = createStatusMap(from: status)
                
                DispatchQueue.main.async {
                    result(statusMap)
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(code: "PRINT_ERROR", message: error.localizedDescription, details: nil))
                }
            }
        }
    }
    
    private func handleCheckStatus(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any],
              let portName = arguments["portName"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Port name is required", details: nil))
            return
        }
        
        Task {
            do {
                guard let printer = printers[portName] else {
                    throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Printer not connected"])
                }
                
                let status = try await printer.getStatus()
                let statusMap = createStatusMap(from: status)
                
                DispatchQueue.main.async {
                    result(statusMap)
                }
            } catch {
                DispatchQueue.main.async {
                    result(FlutterError(code: "STATUS_ERROR", message: error.localizedDescription, details: nil))
                }
            }
        }
    }
    
    private func createDocumentBuilder(from commands: [[String: Any]]) -> StarXpandCommand.DocumentBuilder {
        let documentBuilder = StarXpandCommand.DocumentBuilder()
        let printerBuilder = StarXpandCommand.PrinterBuilder()
        
        for command in commands {
            if let encoding = command["appendEncoding"] as? String {
                // Handle encoding
                handleEncoding(encoding, printerBuilder: printerBuilder)
            }
            else if let cutPaper = command["appendCutPaper"] as? String {
                let cutType: StarXpandCommand.CutType = cutPaper == "FullCut" ? .full : .partial
                printerBuilder.addCut(type: cutType)
            }
            else if let _ = command["openCashDrawer"] as? Int {
                printerBuilder.addDrawer(channel: .no1)
            }
            else if let bitmap = command["appendBitmap"] as? String {
                // Handle bitmap
                handleBitmap(bitmap, command: command, printerBuilder: printerBuilder)
            }
            else if let bitmapText = command["appendBitmapText"] as? String {
                printerBuilder.addText(bitmapText)
            }
            // Add other command handlers
        }
        
        return documentBuilder.addPrinter(printerBuilder)
    }
    
    private func handleEncoding(_ encoding: String, printerBuilder: StarXpandCommand.PrinterBuilder) {
        let charset: String
        switch encoding {
        case "US-ASCII":
            charset = "US-ASCII"
        case "Windows-1252":
            charset = "Windows-1252"
        case "Shift-JIS":
            charset = "Shift-JIS"
        default:
            charset = "UTF-8"
        }
        printerBuilder.addTextEncoding(charset)
    }
    
    private func handleBitmap(_ path: String, command: [String: Any], printerBuilder: StarXpandCommand.PrinterBuilder) {
        guard let imageData = try? Data(contentsOf: URL(fileURLWithPath: path)) else { return }
        
        let width = command["width"] as? Int ?? 576
        let diffusion = command["diffusion"] as? Bool ?? true
        let alignment = command["alignment"] as? String
        
        let imageAlignment: StarXpandCommand.Alignment
        switch alignment {
        case "Left":
            imageAlignment = .left
        case "Center":
            imageAlignment = .center
        case "Right":
            imageAlignment = .right
        default:
            imageAlignment = .left
        }
        
        printerBuilder.addImage(imageData, width: width, effectiveDots: diffusion)
                     .addAlignment(imageAlignment)
    }
    
    private func createStatusMap(from status: StarPrinterStatus) -> [String: Any] {
        return [
            "offline": !status.isOnline,
            "coverOpen": status.coverOpen,
            "cutterError": status.cutterError,
            "receiptPaperEmpty": status.paperEmpty,
            "overTemp": status.overTemp,
            "isSuccess": status.isOnline
        ]
    }
    
    private func getInterfaceType(from typeString: String) throws -> InterfaceType {
        switch typeString.uppercased() {
        case "ALL":
            return .all
        case "LAN":
            return .lan
        case "BLUETOOTH":
            return .bluetooth
        case "USB":
            return .usb
        default:
            throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Unsupported interface type"])
        }
    }
    
    private func detectInterfaceType(from portName: String) throws -> InterfaceType {
        if portName.hasPrefix("TCP:") {
            return .lan
        } else if portName.hasPrefix("BT:") {
            return .bluetooth
        } else if portName.hasPrefix("USB:") {
            return .usb
        }
        throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Unknown interface type"])
    }
    
    // MARK: - FlutterStreamHandler
    
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }
}

// MARK: - StarPrinterDelegate

extension FlutterStarPrinterPlugin: StarPrinterDelegate {
    public func printerDidChangeStatus(_ printer: StarPrinter) {
        Task {
            do {
                let status = try await printer.getStatus()
                let statusMap = createStatusMap(from: status)
                
                DispatchQueue.main.async { [weak self] in
                    self?.eventSink?(statusMap)
                }
            } catch {
                DispatchQueue.main.async { [weak self] in
                    self?.eventSink?(FlutterError(code: "STATUS_ERROR", 
                                                message: error.localizedDescription, 
                                                details: nil))
                }
            }
        }
    }
}